packages/actor-query-operation-sparql-endpoint/lib/ActorQueryOperationSparqlEndpoint.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  DataSources,
+  IDataSource,
+  KEY_CONTEXT_SOURCE,
+  KEY_CONTEXT_SOURCES,
+} from "@comunica/bus-rdf-resolve-quad-pattern";
+import {Actor, IActorArgs, IActorTest, Mediator} from "@comunica/core";
+import {ActionContext} from "@comunica/core";
+import {IMediatorTypeHttpRequests} from "@comunica/mediatortype-httprequests";
+import {BufferedIterator, TransformIterator} from "asynciterator";
+import {SparqlEndpointFetcher} from "fetch-sparql-endpoint";
+import * as RDF from "rdf-js";
+import {termToString} from "rdf-string";
+import {uniqTerms} from "rdf-terms";
+import {Algebra, Factory, toSparql} from "sparqlalgebrajs";
+
+/**
+ * A comunica SPARQL Endpoint Query Operation Actor.
+ */
+export class ActorQueryOperationSparqlEndpoint extends ActorQueryOperation {
+
+  protected static readonly FACTORY: Factory = new Factory();
+  private static ALGEBRA_TYPES: string[] = Object.keys(Algebra.types).map((key) => (<any> Algebra.types)[key]);
+
+  public readonly mediatorHttp: Mediator<Actor<IActionHttp, IActorTest, IActorHttpOutput>,
+    IActionHttp, IActorTest, IActorHttpOutput>;
+  public readonly endpointFetcher: SparqlEndpointFetcher;
+
+  constructor(args: IActorQueryOperationSparqlEndpointArgs) {
+    super(args);
+    this.endpointFetcher = new SparqlEndpointFetcher({
+      fetch: (input?: Request | string, init?: RequestInit) => this.mediatorHttp.mediate({ input, init }),
+      prefixVariableQuestionMark: true,
+    });
+  }
+
+  /**
+   * Recursively find all variables in an operation.
+   * @param {Operation} operation An operation.
+   * @return The variables in the operation.
+   */
+  public static getVariables(operation: Algebra.Operation): RDF.Variable[] {
+    const variables: RDF.Variable[] = [];
+    getVariableCb(operation,  (variable: RDF.Variable) => variables.push(variable));
+    return uniqTerms(variables);
+
+    function getVariableCb(op: Algebra.Operation, variablesCb: (variable: RDF.Variable) => void) {
+      for (const key of Object.keys(op)) {
+        if (Array.isArray(op[key])) {
+          for (const subOperation of op[key]) {
+            getVariableCb(subOperation, variablesCb);
+          }
+        } else if (ActorQueryOperationSparqlEndpoint.ALGEBRA_TYPES.indexOf(op[key].type) >= 0) {
+          getVariableCb(op[key], variablesCb);
+        } else if (op[key].termType === 'Variable') {
+          variablesCb(op[key]);
+        }
+      }
+    }
+  }
+
+  /**
+   * Wrap a pattern in a select query.
+   * @param {Operation} operation An operation.
+   * @return {Project} A select query.
+   */
+  public static patternToSelectQuery(operation: Algebra.Operation): Algebra.Project {
+    if (operation.type === 'project') {
+      return <Algebra.Project> operation;
+    }
+    const variables: RDF.Variable[] = ActorQueryOperationSparqlEndpoint.getVariables(operation);
+    return ActorQueryOperationSparqlEndpoint.FACTORY.createProject(operation, variables);
```

Comment:
> A cheaper solution could be to create a project with an empty list of variables

It's not only needed there. We also need the array of variables in the return object of `run`, so we really need the explicit array.

> The current solution might actually give wrong results with subqueries.

Can you elaborate?

> But then also, when does it occur that this actor gets called with only part of a query?

With SERVICE clauses.

Reply:
Ok, I get it now. There will indeed be some cases where the current code will produce an incorrect list of variables.

> correct solution could be to do the conversion to sparql.js, add the *, and then convert back to sparql algebra :D

Agreed, this would be a bit silly, we can do better than that :-)

> Specifically in the algebra I would have to look up which functions contain variables that should not be used.

Adding functionality into sparqlalgebra would definitely be useful. That would definitely make more sense than adding it into comunica.

Reply:
SparqlAlgebra.js has a inScopeVariables function now

Reply:
Ok, great!

Will you look into applying it? Or shall I?

Reply:
Yea I can add it, shouldn't be much of an issue (assuming I managed to publish the correct files this time)

Reply:
> assuming I managed to publish the correct files this time

It seems to be ok now, I am already using the new version in other packages :-)

Reply:
@joachimvh Can you apply the changes from #200 and #199 as well then in this PR? (but with `^1.0.1` as version)